### ISSUE TITLE
Feature: Deal with Read-Once SFTP servers

### ISSUE THREAD

**Issue by eikenb** [pkg/sftp#345]
There is a type of SFTP server that is termed a "mailbox" style, where you can only download/read each file one time. The current library doesn't work well with these servers as fstat/stat calls are made in the downloading code and that triggers the read and deletes the file before you can actually download it.

A couple related issues in this project: #339  #206 
Here's where Paramiko project (python ssh/sftp lib) dealt with it: https://github.com/paramiko/paramiko/issues/194

The main (only?) SFTP server that behaves like this is IBM Sterling B2B Integrator (SB2Bi). Here is their [troubleshooting page](https://www.ibm.com/support/pages/troubleshooting-guide-why-does-sb2bi-sftp-server-adapter-reports-message-not-extractable-error-when-extractabilitycount1).

The symptom is generally getting a "File does not exist" error when downloading the file as the stat call is interpreted as a read and the file gets deleted before the actual download occurs.

The user (@newdayrising) from #339 dealt with this by writing a custom `WriteTo` method that takes the size of the file as an argument. This works as you are able to call `stat` on a file (on one of these servers) as many times as you like **before** you open the file for reading, but not after.

We should document this and possibly come up with a better way to deal with this situation as these servers seem to be fairly common.

**Comment by ncw** [comment]
I was looking at this as it came up in the rclone forum again: https://forum.rclone.org/t/sftp-fails-to-sync-to-local-failed-to-copy-file-does-not-exist/21759

I thought of two possible solutions...

One would be to put an optional method on the file handle `Size()` which you could type assert and then call to read a cached size. This would be reasonably easy to implement.

However the other was not to read the file size at all.  Looking at the code for `WriteTo` the filesize is used here

https://github.com/pkg/sftp/blob/1cc6c514bddd39dfaa59e97e6e45382d29905fff/client.go#L1053-L1055

And here

https://github.com/pkg/sftp/blob/1cc6c514bddd39dfaa59e97e6e45382d29905fff/client.go#L1098-L1103

Both of these are about closing the window down towards the end of the file, not about detecting the actual end of the file.

So I was wondering whether we could do this a different way without knowing the size of the file in advance?

**Comment by puellanivis** [comment]
So, moving to the sequential version, yeah. We could just totally get rid of the bads by doing a sequential read. It would only be the cases where someone wants to do a parallel `WriteTo` that could then break.

Basically, people would have to opt into the behavior.

**Comment by ncw** [comment]
> So, moving to the sequential version, yeah. We could just totally get rid of the bads by doing a sequential read. It would only be the cases where someone wants to do a parallel `WriteTo` that could then break.
> 
> Basically, people would have to opt into the behavior.

That would work for me - make a nice simple sequential WriteTo which gets used when an option is set.

> P.S.: I think you have now also noticed the very complex code in `WriteTo` which is making its functionality quite difficult to understand. I’m still not wholly clear on just how concurrently any of it runs.

Err, yes! It is the kind of code which you can write, but once written you are now no longer clever enough to debug!

**Comment by puellanivis** [comment]
I’m leaning now towards needing to set an option to get the parallel transfer… the concerns and corner-cases involved are much greater than the presumed simple case. We should shoot for correct naïve behavior, and then allow optimizations that carry special considerations as a feature.

> Err, yes! It is the kind of code which you can write, but once written you are now no longer clever enough to debug!

I remember someone saying once, “this code is very clever, and that’s not a good thing.” 😅

**Comment by drakkan** [comment]
Hi,

I have no way to test with a read once server. I think something like this, while not the ideal solution, should work:

```
diff --git a/client.go b/client.go
index 11bb0bc..d417e27 100644
--- a/client.go
+++ b/client.go
@@ -112,6 +112,19 @@ func UseConcurrentWrites(value bool) ClientOption {
        }
 }
 
+// DisableConcurrentReads disallows the Client to perform concurrent Reads.
+//
+// WriteTo uses Stat/Fstat to get the file size and determines how many concurrent
+// workers to use. This stat call create issues on some read once servers that
+// delete the file. You can disable concurrent reads for these servers.
+// Disabling concurrent reads will degrade performance.
+func DisableConcurrentReads(value bool) ClientOption {
+       return func(c *Client) error {
+               c.disableConcurrentReads = value
+               return nil
+       }
+}
+
 // UseFstat sets whether to use Fstat or Stat when File.WriteTo is called
 // (usually when copying files).
 // Some servers limit the amount of open files and calling Stat after opening
@@ -152,8 +165,9 @@ type Client struct {
 
        // write concurrency is… error prone.
        // Default behavior should be to not use it.
-       useConcurrentWrites bool
-       useFstat            bool
+       useConcurrentWrites    bool
+       useFstat               bool
+       disableConcurrentReads bool
 }
 
 // NewClient creates a new SFTP client on conn, using zero or more option
@@ -1098,6 +1112,10 @@ func (f *File) WriteTo(w io.Writer) (written int64, err error) {
        f.mu.Lock()
        defer f.mu.Unlock()
 
+       if f.c.disableConcurrentReads {
+               return f.writeToSequential(w)
+       }
+
        // For concurrency, we want to guess how many concurrent workers we should use.
        var fileSize uint64
        if f.c.useFstat {
```

@puellanivis do you have better ideas?

**Comment by puellanivis** [comment]
Since the option itself takes a `value` argument, it is not necessary to phrase the option as `DisableXYZ`. We can name it `UseConcurrentReads(value bool)` and note that the default behavior is to use concurrent reads.

Then, under the hood, it can stay `disableConcurrentReads` so that the zero value is still the default. (Design principle of hiding internal complexity behind a simpler/consistent interface.)

**Comment by drakkan** [comment]
> Since the option itself takes a `value` argument, it is not necessary to phrase the option as `DisableXYZ`. We can name it `UseConcurrentReads(value bool)` and note that the default behavior is to use concurrent reads.
> 
> Then, under the hood, it can stay `disableConcurrentReads` so that the zero value is still the default. (Design principle of hiding internal complexity behind a simpler/consistent interface.)
> 
> P.S.: an additional comment should probably note that if concurrent reads are disabled, then the `UseFstat` setting is irrelevant.

Ok, thank you, I'll send a PR later today

**Comment by puellanivis** [comment]
Also, I think we would also want to disable the concurrent reads for `ReadAt` just FYI, not only for `WriteTo`.

You’ll notice that the `f.c.useConcurrentWrites` flag is checked in both `WriteAt` as well as `ReadFrom`.

An overly(?) communicative comment could also note that usually concurrent reads are safer than concurrent writes, so that’s why the default is enabled.

**Comment by drakkan** [comment]
concurrent reads inside `ReadAt` should not create issues to read once servers, we can disable them for uniformity or we can use another name for this new option.

**Comment by puellanivis** [comment]
I have no problems with disabling concurrent reads for `ReadAt` when we disable concurrent reads. At worst, the program does not use the `ReadAt` in this case at all and it doesn’t matter, but best case, we’re consistent in our behavior at least. And as mentioned, it probably won’t matter anyways as it will only happen with buffers around > 32k.

**Comment by drakkan** [comment]
@ncw can you please confirm that #408 fixes this issue? Thank you

**Comment by ncw** [comment]
@darkan

I've reviewed the code - looks good to me.

I've tested it locally - it sure makes things slower so I know it is doing something!

I've added it to rclone in this issue https://github.com/rclone/rclone/issues/5077 and asked for testers. Unfortunately I don't have a machine I can test directly on so I'm reliant on testers pitching in.

**Comment by drakkan** [comment]
Another user confirmed that the patch works with a read once server. This issue should be finally fixed

**Comment by ncw** [comment]
Thank you @drakkan :-)  I'll merge this fix into rclone too for v1.55
